gap/attributes/homomorph.gi

```diff
@@ -0,0 +1,495 @@
+    ErrorNoReturn("the 2nd argument is not a subset of the range of the ",
+                  "1st argument (semigroup homom. by images)");
+  fi;
+  if not IsSubsetSet(AsList(ImagesSource(hom)), elms) then
+    ErrorNoReturn("the 2nd argument is not a subset of the image of the ",
+                  "source of the 1st argument (semigroup homom. by images)");
+  fi;
+  preim := [];
+  for y in elms do
+    Append(preim, PreImagesElm(hom, y));
+  od;
+  return preim;
+end);
+
+InstallMethod(KernelOfSemigroupHomomorphism, "for a semigroup homomorphism",
+[IsSemigroupHomomorphismByImagesOrFunction],
+function(hom)
+  local S, cong, enum, x, y, pairs, i, j;
+
+  S := Source(hom);
+  if IsBijective(hom) then
+    return SemigroupCongruence(S, []);
+  elif Size(ImagesSource(hom)) = 1 then
+    return UniversalSemigroupCongruence(S);
+  fi;
+
+  cong := SemigroupCongruence(S, []);
+  enum := EnumeratorCanonical(S);
+  for i in [1 .. Size(S) - 1] do
+    x := enum[i];
+    for j in [i + 1 .. Size(S)] do
+      y := enum[j];
+      if x ^ hom = y ^ hom then
+        if not [x, y] in cong then
+          pairs := ShallowCopy(GeneratingPairsOfSemigroupCongruence(cong));
+          Add(pairs, [x, y]);
+          cong := SemigroupCongruence(S, pairs);
+          if NrEquivalenceClasses(cong) = Size(ImagesSource(hom)) then
+            return cong;
+          fi;
+        fi;
+      fi;
+    od;
+  od;
+  return cong;
+end);
+
+InstallMethod(String, "for a semigroup homom. by images",
+  [IsSemigroupHomomorphismByImages],
+function(hom)
+  local mapi;
+  mapi := MappingGeneratorsImages(hom);
+  return Concatenation("SemigroupHomomorphismByImages( ",
+          String(Source(hom)), ", ", String(Range(hom)), ", ",
+          String(mapi[1]), ", ", String(mapi[2]), " )");
+end);
+
+InstallMethod(PrintObj, "for a semigroup homom. by images",
+  [IsSemigroupHomomorphismByImages],
+function(hom)
+  Print(String(hom));
+  return;
+end);
+
+InstallMethod(String, "for a semigroup isom. by images",
+  [IsSemigroupHomomorphismByImages and IsBijective],
+function(iso)
+  local mapi;
+  mapi := MappingGeneratorsImages(iso);
+  return Concatenation("SemigroupIsomorphismByImages( ",
+          String(Source(iso)), ", ", String(Range(iso)), ", ",
+          String(mapi[1]), ", ", String(mapi[2]), " )");
+end);
+
+InstallMethod(\=, "compare homom. by images", IsIdenticalObj,
+    [IsSemigroupHomomorphismByImages, IsSemigroupHomomorphismByImages],
+function(hom1, hom2)
+  local i;
+  if Source(hom1) <> Source(hom2)
+      or Range(hom1) <> Range(hom2)
+      or PreImagesRange(hom1) <> PreImagesRange(hom2)
+      or ImagesSource(hom1) <> ImagesSource(hom2) then
+        return false;
+  fi;
+  hom1 := MappingGeneratorsImages(hom1);
+  return hom1[2] = List(hom1[1], i -> ImageElm(hom2, i));
+end);
+
+InstallMethod(ViewObj, "for SHBI/SHBF",
+    [IsSemigroupHomomorphismByImagesOrFunction], 2,
+function(hom)
+  ViewObj(Source(hom));
+  Print(" -> ");
+  ViewObj(Range(hom));
```

Comment:
```suggestion
  Print("\>");
  ViewObj(Source(hom));
  Print("\< \>->\< \>");
  ViewObj(Range(hom));
  Print("\<");
```

Reply:
This is a very soft suggestion, I don't know if this will lead to any better printing or not, the idea is to break the line at `->` rather than anywhere else if possible, but I haven't tested it, and it might not work as intended!

Reply:
It seems to work! Are `\>` and `\>` basically zero-width spaces?